programs/vote/src/vote_state/mod.rs

```diff
@@ -821,12 +831,23 @@ impl VoteState {
         let mut current_vote_state_index = 0;
         let mut new_vote_state_index = 0;
 
+        // Count the number of slots at and before the new root within the current vote state lockouts.  Start with 1
+        // for the new root.  The purpose of this is to know how many slots were rooted by this state update:
+        // - The new root was rooted
+        // - As were any slots that were in the current state but are not in the new state.  The only slots which
+        //   can be in this set are those oldest slots in the current vote state that are not present in the
+        //   new vote state; these have been "popped off the back" of the tower and thus represent finalized slots
```

Comment:
Not necessarily true, the old state could have contained shallow votes on a wrong fork that were later popped off due to lockout expiry (and a fork switch). Given sufficient time this new vote state could have a root that is later than these popped off votes. I believe we'd want to check if these popped off votes are a part of slot hashes as well (checking that they are on the same fork).